### ISSUE TITLE
API: property_id required

### ISSUE THREAD

**Issue by styts** [omeka/omeka-s#1171]
Following the example in the docs, it is possible to create items via the API with this POST body:
```js
{
  "dcterms:title":[
    {
      "type": "literal",
      "property_id": 1,
      "@value": "My item title"
    }
  ]
}
```
Contrary to what one might expect, the `property_id` here is required; without it the property does not get added. This means the the API consumer should know the internal ids of the properties. It would be convenient, if the linking would happen internally, and `property_id` could be omitted.

Additionally, the `dcterms:title` key seems ignored, anything could stand there (since the property_id is being looked at).

It would be great if I could use both the `dcterms:title` as well as the full property uri `http://purl.org/dc/terms/title` as the key, and not the internal property id.

**Comment by Daniel-KM** [comment]
The same issue exist in the url for advanced search, that uses the id, not the properties.

**Comment by zerocrates** [comment]
As you've noticed, the API creates JSON-LD compatible output but requires S-specific json-only keys to do certain things, and properties/values for resources is a big one.

So, the current API doesn't parse the incoming JSON as JSON-LD at all: one major consequence is that the context is ignored. I don't think there's any point in going for any half-measures like recognizing `dcterms:title` divorced from the context based on the conventional or server-side configured prefixes.

So, doing this would require actually tossing the incoming data, or at least the subset of it that represents "real" values, to a JSON-LD processor (probably EasyRDF since we've got that already, but also potentially just the underlying JSON-LD library), resolving to full property URIs and then trying to match those against the currently registered properties. That also implies the ability to download linked contexts, which should be handled for us, but adds a little complexity.

A bit of a complication is that we currently use the presence of `property_id` to determine which things are actually intended to be values, and dropping that for greedy matching against the registered properties could lead to over-matching. There's an in-between possibility of some sort of "auto" value for property_id that would still serve the purpose of declaring "this is a value" but would instruct the API to look up the ID based on the actual JSON-LD parsing.

The aspect Daniel mentions about our internal use of IDs in some places like searching is easier to deal with, and probably worth doing regardless.

**Comment by ptsefton** [comment]
An api endpoint that would take a fully qualified property URI and return a property ID would be really useful. Is there one?

**Comment by styts** [comment]
> An api endpoint that would take a fully qualified property URI and return a property ID would be really useful. Is there one?

As a workaround, I ended up grabbing all the properties and matching up to the `property_id` on the client side: https://github.com/tropy/tropy-omeka/blob/master/src/api.js#L88-L101. Might help you until this is improved.

**Comment by zerocrates** [comment]
@ptsefton You can currently use the `properties` API endpoint to look up properties based on their prefix and local name with the GET parameter `term` (so `?term=dcterms:title`). We don't accept the fully-qualified URI there currently but we could add that without much trouble.

**Comment by ptsefton** [comment]
Thanks @zerocrates that works. Specifically: 

http://localhost/api/properties?term=dcterms%3Acreator&key_identity=your-id&key_credential=your-key

Returns

[{"@context":"http:\/\/localhost\/api-context","@id":"http:\/\/localhost\/api\/properties\/2","@type":"o:Property","o:id":2,"o:local_name":"creator","o:label":"Creator","o:comment":"An entity primarily responsible for making the resource.","o:term":"dcterms:creator","o:vocabulary":{"@id":"http:\/\/localhost\/api\/vocabularies\/1","o:id":1}}]

**Comment by ptsefton** [comment]
In the short term it would be worth explaining this in the API docs, spelling out that the terms used in the JSON examples are placeholders and it's o:id that matters, and how to find it. (Adding an explanation about the key credentials would also be useful.

**Comment by ptsefton** [comment]
In a related question, how does resource_class / @type work. I see you can look up resource_classes via the API, but how do I use that to set the type on an item I want to post to the API?

**Comment by zerocrates** [comment]
It's another one of these things where we're _publishing_ the JSON-LD standard way but _reading_ a parallel key of our own, in this case it's `o:resource_class[o:id]`.

**Comment by jimsafley** [comment]
This is addressed in #1882. Omeka will now look out to the term if the value's `property_id` is "auto".
